### DISCUSSION TITLE
Module-based web app: Custom startup pipeline not supported

### DISCUSSION THREAD

**Discussion by maelscuttle** [dotnet/aspnetcore#26734] [General]
In order to better separate application concerns I tried to introduce a module-based approach for configuring the web-host and the application pipeline, however I am failing to understand what "magic conventions" or restrictions are at work within mostly _Microsoft.AspNetCore.Hosting_ since nothing I try seems to work.

I created a new interface like this:

    public interface IWebModule {
        void ConfigureServices(IServiceCollection services);
        void Configure(IApplicationBuilder app);
    }

At first I included the following into a custom convention-based startup:

    class MyStartup {
        public MyStartup(IEnumerable<IWebModule> modules) { 
            //.ctor stuff
        }
    }

This obviously didn't work since there is no real `IServiceProvider` yet available. It seems there is only a pseudo-service locator within [GenericWebHostBuilder](https://github.com/dotnet/aspnetcore/blob/b7b300143518f86d5e4ff6630688c7d578efda07/src/Hosting/Hosting/src/GenericHost/GenericWebHostBuilder.cs#L380).
Instead of relying on `WebHostBuilder.UseStartup<T>()`, I then tried to utilize `WebHostBuilder.Configure(Action<IApplicationBuilder>)` and call the following:

    builder.Configure(app => {
        var modules = app.ApplicationServices.GetServices<IWebModule>();
        foreach(var module) {
            // service reg left out for readability
            module.Configure(app);
        }));

While just calling `builder.Configure()` works, this one then throws exception such as: `InvalidOperationException: Unable to resolve service for type 'Microsoft.AspNetCore.Mvc.ApplicationParts.ApplicationPartManager' while attempting to activate 'Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation.RuntimeViewCompilerProvider'.`

Next I tried again with a custom startup implementing `IStartup,` but resolving services beforehand.

    builder.ConfigureServices(services => {
        services.AddSingleton<IStartup>(provider =>
            var modules = provider.GetServices<IWebModule>();
            return new MyStartup(modules);
        });
    }
    // should technically create a new DelegateStartup which resolves IStartup through the service provider
    builder.Configure(app => {});

Which then results in a `NotSupportedException` telling that `IStartup` isn't supported.

Last I did was trying to create my own `Configure()`-Extension, but that didn't work either, since `Configure` seems to internally cast the `IWebHostBuilder` to `ISupportsStartup`, which is internal.

Is there a better approach to accomplish this? Or did I completely misunderstand something here? @davidfowl @Tratcher

**Comment by Tratcher** [comment]
The basic problem here is that while multiple calls to ConfigureServices are supported, only one call to Configure is supported. Services are much less order sensitive, but Configure is responsible for building an order sensitive pipeline.

Your best approach was probably:
```
builder.Configure(app => {
    var modules = app.ApplicationServices.GetServices<IWebModule>();
    foreach(var module) {
        // service reg left out for readability
        module.Configure(app);
    }));
```

The MVC errors don't seem to be an issue with your approach, only that you're missing some services that need to be registered. Admittedly IWebModule can't register additional services. The only module system designed for that is HostingStartup. https://docs.microsoft.com/en-us/aspnet/core/fundamentals/host/platform-specific-configuration?view=aspnetcore-3.1

**Comment by maelscuttle** [comment]
@Tratcher 

Thank you. I wasn't aware of `HostingStartup`, which however doesn't fit my use case as it works on an assembly level, not a type level it seems.

I'm not really trying to call `Configure` multiple times. What I am essentially trying to do is configure both the `IServiceCollection` as well as the `IApplicationBuilder` using a type-based module approach, so I can add entire sets of "features" to my web app (e.g. `IWebHostBuilder.AddModule<SitemapModule>()`) without having to write explicit extension methods, for both the services and the app builder, and also that I can scan for them by convention within an assembly.

What I kinda fail to understand is why my modules technically can't register additional services. I know, the service provider is already built when I would resolve them, so that doesn't make much sense, however when browsing through the AspNetCore hosting code (especially the `GenericWebHostBuilder`), it seems to me that when using `IWebHostBuilder.Configure()`, it internally registers a `IStartup` (`DelegateStartup`) which is then resolved using an internal `IServiceProvider`, that THEN calls `ConfigureServices`, and later on a new `IServiceProvider` is being built. So that could technically work?

My MVC errors directly relate to that issue. The exception occurs when I try to resolve my modules from `IApplicationBuilder.ApplicationServices`, because I'm using RazorPages and I haven't called `IServiceCollection.AddRazorPages()` yet. The reason: I'd like to have a `RazorPagesWebModule` like this:

    public class RazorPagesWebModule : IWebModule {

        public void ConfigureServices(IServiceCollection services) {
            services.AddRazorPages();
        }

        public void Configure(IApplicationBuilder app) {
        
        }
    }

But I feel like something like that just isn't possible?

**Reply by Tratcher** [reply]
Don't pay too much attention to what's going on inside WebHostBuilder. Yes that code did have two service containers, but it caused multiple issues with service lifetimes and we had to deprecate that model. The new model in 3.0 uses HostBuilder and a single service container. Startup is no longer resolved from the container, it's activated directly using only services that the host owns (IConfiguration, IHostEnvironment, etc.).

Because there's only one container now you can't register services, resolve them, and then register additional ones. To build a dynamic module system like this you need to provide your own container or other way of discovering, registering, and activating the modules in advance such that they can then register their own services on the apps container before it gets built.

**Comment by davidfowl** [comment]
Orchard supports this natively, I'd recommend starting there 
cc @sebastienros
